src/github.com/couchbase/sync_gateway/rest/handler.go

```diff
@@ -515,6 +517,100 @@ func (h *handler) writeStatus(status int, message string) {
 	h.response.Write(jsonOut)
 }
 
+var kRangeRegex = regexp.MustCompile("^bytes=(\\d+)?-(\\d+)?$")
+
+// Detects and partially HTTP content range requests.
+// If the request _can_ accept ranges, sets the "Accept-Ranges" response header to "bytes".
+//
+// If there is no Range: request header, or if its valid is invalid, returns a status of 200,
+// meaning that the caller should return the entire response as usual.
+//
+// If there is a request range but it exceeds the contentLength, returns status 416. The caller
+// should treat this as an error and abort, returning that HTTP status code.
+//
+// If there is a useable range, it returns status 206 and the start and end in Go slice terms, i.e.
+// starting at 0 and with the end non-inclusive. It also adds a "Content-Range" response header.
+// It is then the _caller's_ responsibility to set it as the response status code (by calling
+// h.response.WriteHeader(status)), and then write the indicated subrange of the response data.
+func (h *handler) handleRange(contentLength uint64) (status int, start uint64, end uint64) {
+	status = http.StatusOK
+	if h.rq.Method == "GET" || h.rq.Method == "HEAD" {
+		h.setHeader("Accept-Ranges", "bytes")
+		status, start, end = parseHTTPRangeHeader(h.rq.Header.Get("Range"), contentLength)
+		if status == http.StatusPartialContent {
+			h.setHeader("Content-Range", fmt.Sprintf("bytes %d-%d/%d", start, end, contentLength))
+			h.setStatus(http.StatusPartialContent, "Partial Content")
+			end += 1 // make end non-inclusive, as in Go slices
+		}
+	}
+	return
+}
+
+// Given an HTTP "Range:" header value, parses it and returns the approppriate HTTP status code,
+// and the numeric byte range if appropriate:
+// * If the Range header is empty or syntactically invalid, it ignores it and returns status=200.
+// * If the header is valid but exceeds the contentLength, it returns status=416 (Not Satisfiable).
+// * Otherwise it returns status=206 and sets the start and end values in HTTP terms, i.e. with
+//   the first byte numbered 0, and the end value inclusive (so the first 100 bytes are 0-99.)
+func parseHTTPRangeHeader(rangeStr string, contentLength uint64) (status int, start uint64, end uint64) {
+	// http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.35
```

Comment:
rfc2616 includes the ability to specify multiple ranges (e.g. "bytes=500-600,800-900").  These would just get rejected as not matching the regex, correct?  I assume we don't intend to support that type of range request (yet?).